kong/tools/stream_api.lua

```diff
@@ -5,114 +5,355 @@
 
 local lpack = require "lua_pack"
 
+local kong = kong
+local st_pack = lpack.pack
+local st_unpack = lpack.unpack
+local concat = table.concat
+local assert = assert
+local type = type
+local tostring = tostring
+local tcp = ngx.socket.tcp
+local req_socket = ngx.req.socket
+local exit = ngx.exit
+local log = ngx.log
+local ERR = ngx.ERR
+local DEBUG = ngx.DEBUG
+local WARN = ngx.WARN
+local exiting = ngx.worker.exiting
 
-local kong       = kong
-local st_pack    = lpack.pack
-local st_unpack  = lpack.unpack
-local st_format  = string.format
-local table_concat = table.concat
-local assert     = assert
+local CLOSE = 444
+local OK = 0
 
-local MAX_DATA_LEN = 8000
-local PREFIX = ngx.config.prefix()
+
+local PACK_F = "=CI"
+
+-- unsigned char length
+local MAX_KEY_LEN = 2^8 - 1
+
+-- since the length is represented by an unsigned int we could theoretically
+-- go up to 2^32, but that seems way beyond the amount of data that we should
+-- expect to see exchanged over this interface
+local MAX_DATA_LEN = 2^22 - 1
+
+local HEADER_LEN = #st_pack(PACK_F, MAX_KEY_LEN, MAX_DATA_LEN)
+
+local SOCKET_PATH = "unix:" .. ngx.config.prefix() .. "/stream_rpc.sock"
 
 local stream_api = {}
 
 local _handlers  = {}
 
 
+-- # RPC format
```

Comment:
Is it possible to reuse wrpc encoding (and even some codes)? We already have lots of adhoc RPC's in Kong, unifying them will be a huge win from my view.

Reply:
it might be possible, but i doubt it would be approriate. wrpc has a lot of code in order to multiplex services and share a single connection to use from any worker.  this is ok to reduce the latency impact from long distance internet connections, but just overhead when used on unix sockets.

I was thinking the exact opposite:  wrpc requires some message passing between workers.  right now it uses a very ad-hoc facility in `channels.lua` built on top of shared dictionaries, but i'd like to move that to something else.  maybe this thing, or maybe the new work in worker events.

Reply:
Oh yes that reminds me of the eventing system that @chronolaw is been building, maybe this PR and the wrpc side can indeed be made to depend on that in the future. Anyway, @javierguerragiraldez your comment makes sense to me, so consider this thread resolved and no action needed.

Reply:
I agree, having a one-off RPC encoding is not great. The majority of this code was actually written ~6 months ago, so wRPC just wasn't on my radar back then.

I think my main worry is that the additional overhead of wRPC would eat into performance when sending large payloads (maybe this is unfounded though?).

The more that I think about it, given that the sole use-case for this API is exchanging prometheus metric data (raw text that is fed directly to a downstream prometheus collector HTTP client), I'd actually argue that this API could be dumbed down and made into a plaintext, line-based protocol (`${request_key_or_status_code}\r\n${content_length}\r\n${payload}`). I know this is moving in the opposite direction of your suggestion, but the benefit is that it _would_ simplify the code _and_ let us drop the lua_pack dependency. Good idea/bad idea?